hashicorp/terraform-provider-tfe
# Integrate terraform-plugin-framework, and rewrite the tfe_variable resource

## Description

Fixes #839.

Wow, lots of sturm und drang for one little bug fix, huh? But anyway, this starts us on the path to adopting the shiny new plugin framework _(which, by the way, is actually incredibly excellent to work with compared to the v2 sdk)_ and validates our theory that it grants the power to resolve that perennially obnoxious "can't tell absent from zero" family of defects.

TODO: 

- [x] Minor docs tweak

## Testing plan

Sorry, this is outrageously convoluted, which is why no one apparently noticed the bug until the reporter on #839.

1.  Make an arbitrary terraform config that relies on a sensitive variable being set to a particular value. (Probably easiest is to just switch between two output values based on an equality check against a known value.)
2. Make a terraform config that uses this provider to create a TFC workspace (which will use the config from step 1), and a sensitive variable in that workspace. The var should have the value set to an **empty string,** and should have `ignore_changes = [value]`. Like this:

    ```hcl
    resource "tfe_variable" "affected" {
      key          = "my_var"
      value        = ""
      category     = "terraform"
      workspace_id = tfe_workspace.var-sensitive-nuke-repro.id
      description  = "not edited yet"
      sensitive    = true
      lifecycle {
        ignore_changes = [value]
      }
    }
    ```
3. Apply the config from step 2.
4. **Manually, in the TFC UI,** change the value of the sensitive variable in your new workspace to the Particular Value that your config from 1 relies on.
5. Do an apply in the new workspace and verify the value is set as expected. 
6. Edit your config from step 2 to change **the `description`** of the variable, leaving `value` unchanged. Do an apply, and note that Terraform only reports a change to description as being necessary. 
7. Do an apply in the new workspace again.
    - In released versions of the provider, this now breaks because the value was **forcibly, silently reset** to an empty string, despite a description edit being the only planned change.
    - With this PR, it works fine.

**Additionally:** Slam on the `tfe_variable` resource as hard as you can and try to squeeze any kind of unexpected behavior out of it. It ought to behave. 

## Output from acceptance tests

```
TF_ACC=1 TF_LOG_SDK_PROTO=OFF go test $(go list ./... |grep -v 'vendor') -v -run TestAccTFEVariable_ -timeout 15m
?   	github.com/hashicorp/terraform-provider-tfe	[no test files]
=== RUN   TestAccTFEVariable_basic
--- PASS: TestAccTFEVariable_basic (37.70s)
=== RUN   TestAccTFEVariable_basic_variable_set
--- PASS: TestAccTFEVariable_basic_variable_set (32.52s)
=== RUN   TestAccTFEVariable_update
--- PASS: TestAccTFEVariable_update (55.84s)
=== RUN   TestAccTFEVariable_update_key_sensitive
--- PASS: TestAccTFEVariable_update_key_sensitive (55.95s)
=== RUN   TestAccTFEVariable_import
--- PASS: TestAccTFEVariable_import (37.09s)
=== RUN   TestAccTFEVariable_rewrite
--- PASS: TestAccTFEVariable_rewrite (56.65s)
PASS
ok  	github.com/hashicorp/terraform-provider-tfe/tfe	275.978s
?   	github.com/hashicorp/terraform-provider-tfe/version	[no test files]
```
